salt/_utils/kubernetes_utils.py

```diff
@@ -0,0 +1,829 @@
+
+try:
+    import kubernetes.config
+    import kubernetes.client as k8s_client
+    import kubernetes.client.apis as k8s_apis
+except ImportError:
+    HAS_LIBS = False
+else:
+    HAS_LIBS = True
+    ALL_APIS = frozenset(
+        api for _, api in inspect.getmembers(k8s_apis, inspect.isclass)
+    )
+
+__virtualname__ = 'metalk8s_kubernetes'
+
+
+def __virtual__():
+    if not HAS_LIBS:
+        return False, 'Missing dependencies: kubernetes'
+
+    return __virtualname__
+
+
+# Roughly equivalent to an Enum, for Python 2
+class ObjectScoped:
+    NAMESPACE = 'namespaced'
+    CLUSTER = 'cluster'
+
+    ALL_VALUES = (NAMESPACE, CLUSTER)
+
+    def __init__(self, value):
+        if value not in self.ALL_VALUES:
+            raise ValueError(
+                'Value must be one of {}.'.format(self.ALL_VALUES)
+            )
+        self.value = value
+
+    def __eq__(self, other):
+        if isinstance(other, ObjectScoped):
+            return self.value == other.value
+        if isinstance(other, six.string_types):
+            return self.value == other
+        return NotImplemented
+
+    def __repr__(self):
+        return 'ObjectScope({})'.format(self.value)
+
+
+class ApiClient(object):
+    CRUD_METHODS = {
+        'create': 'create',
+        'retrieve': 'read',
+        'update': 'patch',
+        'delete': 'delete',
+        'replace': 'replace',
+        'list': 'list',
+    }
+
+    def __init__(self, api_cls, name, all_namespaces_name=None):
+        if api_cls not in ALL_APIS:
+            raise ValueError(
+                '`api_cls` must be an API from `kubernetes.client.apis`'
+            )
+        self._api_cls = api_cls
+        self._name = name
+        self._all_namespaces_name = all_namespaces_name
+        self._api = None
+        self._client = None
+
+        # Attach the API CRUD methods at construction, so we can fail at
+        # import-time instead of runtime
+        self._api_methods = {
+            method: getattr(api_cls, self._method_name(verb))
+            for method, verb in self.CRUD_METHODS.items()
+        }
+
+        if self._all_namespaces_name:
+            self._api_methods['list_all_namespaces'] = \
+                getattr(api_cls, self._all_namespaces_name)
+
+    api_cls = property(operator.attrgetter('_api_cls'))
+    name = property(operator.attrgetter('_name'))
+
+    create = property(lambda self: self._method('create'))
+    retrieve = property(lambda self: self._method('retrieve'))
+    update = property(lambda self: self._method('update'))
+    delete = property(lambda self: self._method('delete'))
+    replace = property(lambda self: self._method('replace'))
+    list = property(lambda self: self._method('list'))
+
+    @property
+    def list_all_namespaces(self):
+        assert self._all_namespaces_name, (
+            'Cannot use "list_all_namespaces" for this client'
+        )
+        return self._method('list_all_namespaces')
```

Comment:
Mmmh, not super fond of this. I'd rather have a tailor-made `list` method that handles this particular case, e.g.:

```python
@property
def list(self):
    def _list(*args, all_namespaces=False, **kwargs):
        if all_namespaces:
            return self._api_methods['list_all_ns'](self.api, *args, **kwargs)
        return self._method('list')(*args, **kwargs)
    return _list
```